helpers/manifest-parser.js

```diff
@@ -0,0 +1,325 @@
+  // var dummy = document.createElement('div');
+  // dummy.style.color = 'white';
+  // dummy.style.color = color.value;
+  // if (dummy.style.color !== 'white') {
+  //   return color;
+  // }
+  // dummy.style.color = 'black';
+  // dummy.style.color = color.value;
+  // if (dummy.style.color !== 'black') {
+  //   return color;
+  // }
+  return color;
+
+  // color.value = undefined;
+  // color.warning = 'ERROR: color parsing failed';
+
+  // return color;
+}
+
+function parseName(jsonInput) {
+  return parseString(jsonInput.name, true);
+}
+
+function parseShortName(jsonInput) {
+  return parseString(jsonInput.short_name, true);
+}
+
+function parseStartUrl(jsonInput) {
+  // TODO: parse url using manifest_url as a base (missing).
+  // start_url must be same-origin as Document of the top-level browsing context.
+  return parseURL(jsonInput.start_url);
+}
+
+function parseDisplay(jsonInput) {
+  let display = parseString(jsonInput.display, true);
+
+  if (display.value && ALLOWED_DISPLAY_VALUES.indexOf(display.value.toLowerCase()) === -1) {
+    display.value = undefined;
+    display.warning = 'ERROR: \'display\' has an invalid value, will be ignored.';
+  }
+
+  return display;
+}
+
+function parseOrientation(jsonInput) {
+  let orientation = parseString(jsonInput.orientation, true);
+
+  if (orientation.value &&
+      ALLOWED_ORIENTATION_VALUES.indexOf(orientation.value.toLowerCase()) === -1) {
+    orientation.value = undefined;
+    orientation.warning = 'ERROR: \'orientation\' has an invalid value, will be ignored.';
+  }
+
+  return orientation;
+}
+
+function parseIcon(raw) {
+  // TODO: pass manifest url as base.
+  let src = parseURL(raw.src);
+  let type = parseString(raw.type, true);
+
+  let density = {
+    raw: raw.density,
+    value: 1,
+    warning: undefined
+  };
+  if (density.raw !== undefined) {
+    density.value = parseFloat(density.raw);
+    if (isNaN(density.value) || !isFinite(density.value) || density.value <= 0) {
+      density.value = 1;
+      density.warning = 'ERROR: icon density cannot be NaN, +∞, or less than or equal to +0.';
+    }
+  }
+
+  let sizes = parseString(raw.sizes);
+  if (sizes.value !== undefined) {
+    let set = new Set();
+    sizes.value.split(/\s/).forEach(size => set.add(size.toLowerCase()));
+
+    sizes.value = set.size > 0 ? set : undefined;
+  }
+
+  return {
+    raw,
+    value: {
+      src,
+      type,
+      density,
+      sizes
+    },
+    warning: undefined
+  };
+}
+
+function parseIcons(jsonInput) {
+  const raw = jsonInput.icons;
+  let value;
+
+  if (raw === undefined) {
+    return {
+      raw,
+      value,
+      warning: undefined
+    };
+  }
+
+  if (!Array.isArray(raw)) {
+    return {
+      raw,
+      value,
+      warning: 'ERROR: \'icons\' expected to be an array but is not.'
+    };
+  }
+
+  // TODO(bckenny): spec says to skip icons missing `src`. Warn instead?
+  value = raw.filter(icon => !!icon.src).map(parseIcon);
```

Comment:
as opposed to invalid values, where the spec says to warn and then set to `undefined`, the spec is very clear that icons without a `src` [should be skipped with no warning](https://www.w3.org/TR/appmanifest/#dfn-steps-for-processing-an-array-of-images). In this case I think we may want to deviate from the spec to warn that there may be an error (why is an icon there?) even though it's acceptable